src/status_im/chat/models/commands.cljs

```diff
@@ -0,0 +1,124 @@
+          (if can-use-for-dapps? 16 0)))
+
+(defn can-be-suggested?
+  [first-char name-key text]
+  (fn [command]
+    (let [name (get command name-key)]
+      (let [text' (cond
+                    (.startsWith text first-char)
+                    text
+
+                    (str/blank? text)
+                    first-char
+
+                    :default
+                    nil)]
+        (.startsWith (str first-char name) text')))))
+
+(defn get-mixable-commands [{:contacts/keys [contacts]}]
+  (->> contacts
+       (vals)
+       (filter :mixable?)
+       (mapv :commands)
+       (mapcat #(into [] %))
+       (reduce (fn [acc [k v]] (update acc k #(into % v))) {})))
+
+(defn get-mixable-identities [{:contacts/keys [contacts]}]
+  (->> contacts
+       (vals)
+       (filter :mixable?)
+       (map (fn [{:keys [whisper-identity]}]
+              {:identity whisper-identity}))))
+
+(defn find-suggestions
+  ([commands text]
+    (find-suggestions chat-consts/command-char :name commands text))
+  ([first-char name-key commands text]
+   (->> commands
+        (map val)
+        (map (fn [items]
+               (filter #((can-be-suggested? first-char name-key text) %) items)))
+        (remove empty?)
+        (flatten))))
+
+(defn get-possible-requests
+  [{:keys [current-chat-id] :as db}]
+  (let [requests (->> (get-in db [:chats current-chat-id :requests])
+                      (map (fn [{:keys [type chat-id bot] :as req}]
+                             [type (map (fn [resp]
+                                          (assoc resp :request req))
+                                        (get-in db [:contacts/contacts (or bot chat-id) :responses type]))]))
+                      (remove (fn [[_ items]] (empty? items)))
+                      (into {}))]
+    (find-suggestions requests "")))
+
+(defn get-possible-commands
+  [{:keys [current-chat-id] :as db}]
+  (->> (get-in db [:chats current-chat-id :contacts])
+       (into (get-mixable-identities db))
+       (map (fn [{:keys [identity]}]
+              (let [commands (get-in db [:contacts/contacts identity :commands])]
+                (find-suggestions commands ""))))
+       (flatten)))
+
+(defn get-possible-global-commands
+  [{:keys [global-commands] :as db}]
+  (find-suggestions chat-consts/bot-char :name global-commands ""))
+
+(defn commands-for-chat
+  [{:keys          [global-commands chats]
+    :contacts/keys [contacts]
+    :accounts/keys [accounts current-account-id]
+    :as            db} chat-id]
+  (let [global-commands (get-possible-global-commands db)
+        commands        (get-possible-commands db)
+        account         (get accounts current-account-id)
+        commands        (-> (into [] global-commands)
+                            (into commands))
+        {chat-contacts :contacts} (get chats chat-id)]
+    (->> commands
+         (remove (fn [{:keys [scope]}]
+                   (or
+                     (and (:registered-only? scope)
```

Comment:
Can we bring out these predicates into fns in let? So it is more clear semantically what conditions must be met. Something like:

```
(or (personal-chat? scope) foo? bar? ...)
```

(Can bind `chat-contacts` with partial fn, for example)

Reply:
I tried this approach and must say that it actually makes things a bit more complicated.
Reasons are the following:
1) We also need to pass `account` and `chat-contacts`, or `db` to these functions.
2) If we create separate functions for scopes, they will look this way:

```
(defn registered-only? [scope account]
  (or (:registered-only? scope)
      (:address account)))
```
As for me, the name doesn't really represent what's happening inside + it becomes difficult to understand the function itself.